File: deob-ast/src/main/java/dev/openrs2/deob/ast/util/NodeUtils.java
package dev.openrs2.deob.ast.util;

import java.util.function.Consumer;

import com.github.javaparser.ast.Node;

public final class NodeUtils {
	@SuppressWarnings("unchecked")
	public static <T extends Node> void walk(Node node, Node.TreeTraversal traversal, Class<T> type, Consumer<T> consumer) {
		node.walk(traversal, n -> {
			if (type.isAssignableFrom(n.getClass())) {
				consumer.accept((T) n);
			}
		});
	}

	private NodeUtils() {
		/* empty */
	}
}

File: deob-ast/src/main/java/dev/openrs2/deob/ast/util/NodeUtils.kt
package dev.openrs2.deob.ast.util

import com.github.javaparser.ast.Node

public inline fun <reified T : Node> Node.walk(crossinline consumer: (T) -> Unit) {
    this.walk(Node.TreeTraversal.POSTORDER) {
        if (it is T) {
            consumer(it)
        }
    }
}
